synapse/cores/storage.py

```diff
@@ -0,0 +1,912 @@
+        link ((str, dict)): Link tufo containing information for creating the Storage object.  This may include path
+                            information, authentication information, etc.
+        **conf (dict):  Additional configible options for the storage layer.
+    '''
+    def __init__(self,
+                 link,
+                 **conf):
+        s_config.Config.__init__(self)
+        if conf:
+            self.setConfOpts(conf)
+
+        #############################################################
+        # buses to save/load *raw* save events
+        #############################################################
+        self.savebus = s_eventbus.EventBus()
+        self.loadbus = s_eventbus.EventBus()
+
+        self._link = link
+
+        self.onfini(self._defaultFiniCoreStor)
+
+        # Various locks
+        self.xlock = threading.Lock()
+        # Transactions are a storage-layer concept
+        self._store_xacts = {}
+
+        # Dicts for storing retrieval methods
+        self.sizebymeths = {}
+        self.rowsbymeths = {}
+
+        # Register handlers for lifting rows/siexes
+        self.initRowsBy('gt', self.rowsByGt)
+        self.initRowsBy('lt', self.rowsByLt)
+        self.initRowsBy('ge', self.rowsByGe)
+        self.initRowsBy('le', self.rowsByLe)
+        self.initRowsBy('range', self.rowsByRange)
+
+        self.initSizeBy('ge', self.sizeByGe)
+        self.initSizeBy('le', self.sizeByLe)
+        self.initSizeBy('range', self.sizeByRange)
+
+        # Provide a dictionary of by / meth values a storage layer
+        # can add too and a Cortex can pull from
+        self.joinsbymeths = [
+            ('le', self.joinsByLe),
+            ('lt', self.joinsByLt),
+            ('gt', self.joinsByGt),
+            ('ge', self.joinsByGe),
+        ]
+
+        # Events for handling savefile loads/saves
+        self.loadbus.on('core:save:add:rows', self._loadAddRows)
+        self.loadbus.on('core:save:del:rows:by:iden', self._loadDelRowsById)
+        self.loadbus.on('core:save:del:rows:by:prop', self._loadDelRowsByProp)
+        self.loadbus.on('core:save:set:rows:by:idprop', self._loadSetRowsByIdProp)
+        self.loadbus.on('core:save:del:rows:by:idprop', self._loadDelRowsByIdProp)
+        self.loadbus.on('syn:core:blob:set', self._onSetBlobValu)
+        self.loadbus.on('syn:core:blob:del', self._onDelBlobValu)
+
+        # Cache blob save mesgs which may be fired during storage layer init
+        _blobMesgCache = []
+        self.savebus.on('syn:core:blob:set', _blobMesgCache.append)
+        self.savebus.on('syn:core:blob:del', _blobMesgCache.append)
+
+        # Perform storage layer initializations
+        self._initCoreStor()
+
+        # Disable the blob message caching
+        self.savebus.off('syn:core:blob:set', _blobMesgCache.append)
+        self.savebus.off('syn:core:blob:del', _blobMesgCache.append)
+
+        # process a savefile/savefd if we have one (but only one of the two)
```

Comment:
capturing sidechannel notes - the savefile mechanism does WORK across all cortex implementations, and has a the second order effect of acting like an audit log of things going into a core and being removed from a core.  By having the generic implementation in the Storage object, storage implementors don't have to worry about implementing it themselves.